src/cargo/ops/cargo_install.rs

```diff
@@ -322,7 +323,43 @@ impl<'cfg, 'a> InstallablePackage<'cfg, 'a> {
             })
             .collect::<CargoResult<_>>()?;
         if binaries.is_empty() {
-            bail!("no binaries are available for install using the selected features");
+            // Cargo already warns the user if they use a target specifier that matches nothing,
+            // but we want to error if the user asked for a _particular_ binary to be installed,
+            // and we didn't end up installing it.
+            //
+            // NOTE: This _should_ be impossible to hit since --bin=does_not_exist will fail on
+            // target selection, and --bin=requires_a without --features=a will fail with "target
+            // .. requires the features ..". But rather than assume that's the case, we define the
+            // behavior for this fallback case as well.
+            if let CompileFilter::Only { bins, examples, .. } = &self.opts.filter {
+                let mut any_specific = false;
+                if let FilterRule::Just(ref v) = bins {
+                    if !v.is_empty() {
+                        any_specific = true;
+                    }
+                }
+                if let FilterRule::Just(ref v) = examples {
+                    if !v.is_empty() {
+                        any_specific = true;
+                    }
+                }
+                if any_specific {
+                    bail!("no binaries are available for install using the selected features");
+                }
+            }
+
+            // If there _are_ binaries available, but none were selected given the current set of
+            // features, let the user know.
+            //
+            // Note that we know at this point that _if_ bins or examples is set to `::Just`,
+            // they're `::Just([])`, which is `FilterRule::none()`.
+            if self.pkg.targets().iter().any(|t| t.is_bin()) {
```

Comment:
Just one corner case and we're good to go!

```suggestion
            if self.pkg.targets().iter().any(|t| t.is_executable()) {
```

Reply:
Ah, good catch!